webapp/src/Controller/API/SubmissionController.php

```diff
@@ -219,23 +318,120 @@ public function addSubmissionAction(Request $request)
+                throw new BadRequestHttpException("The 'files' attribute should be an array with a single item, containing an object with a base64 encoded data field");
+            }
+
+            if (isset($filesList[0]['mime']) && $filesList[0]['mime'] !== 'application/zip') {
+                throw new BadRequestHttpException("The 'files[0].mime' attribute should be application/zip if provided");
+            }
+
+            $data        = $filesList[0]['data'];
+            $decodedData = base64_decode($data);
+            if ($decodedData === false) {
+                throw new BadRequestHttpException("The 'files[0].data' attribute is not base64 encoded");
+            }
+
+            $tmpDir = $this->dj->getDomjudgeTmpDir();
+
+            // Now write the data to a temporary ZIP file
+            if (!($tempZipFile = tempnam($tmpDir, 'submission_zip-'))) {
+                throw new ServiceUnavailableHttpException(null,
+                    sprintf('Could not create temporary file in directory %s', $tmpDir));
+            }
+
+            if (file_put_contents($tempZipFile, $decodedData) === false) {
+                throw new ServiceUnavailableHttpException(
+                    null,
+                    sprintf("Could not write ZIP to temporary file '%s'.", $tempZipFile)
+                );
+            }
+
+            $zip = $this->dj->openZipFile($tempZipFile);
+
+            $files = [];
+            for ($i = 0; $i < $zip->numFiles; $i++) {
+                $name   = $zip->getNameIndex($i);
+                $source = $zip->getFromIndex($i);
+
+                if (!($tempFileName = tempnam($tmpDir, 'submission-'))) {
+                    throw new ServiceUnavailableHttpException(null,
+                        sprintf('Could not create temporary file in directory %s', $tmpDir));
+                }
+                if (file_put_contents($tempFileName, $source) === false) {
+                    throw new ServiceUnavailableHttpException(
+                        null,
+                        sprintf("Could not write to temporary file '%s'.", $tempFileName)
+                    );
+                }
+                $files[]     = new UploadedFile($tempFileName, $name, null, null, true);
+                $tempFiles[] = $tempFileName;
+            }
+
+            $zip->close();
+            unlink($tempZipFile);
+        } else {
+            // Files are uploaded as actual files, get them
+            $files = $request->files->get('code') ?: [];
+            if (!is_array($files)) {
+                $files = [$files];
+            }
         }
 
         // Now submit the solution
-        $team       = $this->dj->getUser()->getTeam();
         $submission = $this->submissionService->submitSolution(
             $team, $problem, $problem->getContest(), $language,
-            $files, null, null, $entryPoint, null, null, $message
+            $files, null, null, $entryPoint, $submissionId, $time, $message
```

Comment:
So the supplied ID we store as an external ID. Is this compliant with the spec? Especially the over-designed "proxy accepts submissions for later forwarding" use case?

Reply:
That is exactly the reason we need this. What we do is when this is merged, we return the external ID if it is set, which the proxy can send to us.